tensorflow/python/kernel_tests/conv_ops_test.py

```diff
@@ -3266,6 +3267,164 @@ def Test(self):
   return Test
 
 
+class FusedConv2DTest(test.TestCase):
+
+  def _CreateNumpyTensor(self, shape):
+    total_size = 1
+    for s in shape:
+      total_size *= s
+    return np.arange(1, total_size + 1, dtype=np.float32).reshape(shape)
+
+  def _CreateConv2D(self, input_values, filters,
+                    strides=[1, 1], padding="SAME"):
+    return nn_ops.convolution(
+              input_values,
+              filters,
+              strides=strides,
+              padding=padding)
+
+  @test_util.deprecated_graph_mode_only
+  def testAddWithRefCountOne(self):
+    expected_output = [
+        113377, 125570, 77305, 86738, 19433, 22226, 60681,
+        70722, 36291, 43718, 7143, 9206, 9785, 12098,
+        4783, 6366, 779, 1134]
+    tensor_in_sizes = [1, 3, 3, 2]
+    filter_in_sizes = [2, 2, 2, 2]
+    bias_in_sizes = [2]
+
+    x = self._CreateNumpyTensor(tensor_in_sizes)
+    filter_in = self._CreateNumpyTensor(filter_in_sizes)
+    bias_in = self._CreateNumpyTensor(bias_in_sizes)
+    # To get different weights for filter
+    ofs = 1
+
+    conv1 = self._CreateConv2D(x, filter_in)
+    conv2 = self._CreateConv2D(conv1, filter_in + ofs)
+
+    conv = self._CreateConv2D(conv1, filter_in - ofs)
+    bias_add = nn_ops.bias_add(conv, bias_in)
+    add = math_ops.add_n([bias_add, conv2])
+
+    self.assertAllEqual(
+        np.rint(expected_output),
+        self.evaluate(add).reshape(-1))
+
+  @test_util.deprecated_graph_mode_only
+  def testAddWithRefCountTwoAndRunAddLast(self):
+    expected_output = [
+        1.907175e+06, 2.253505e+06, 7.809210e+05, 9.537180e+05,
+        1.184170e+05, 1.523070e+05, 5.367010e+05, 6.803700e+05,
+        1.867090e+05, 2.529460e+05, 2.362300e+04, 3.522600e+04,
+        5.121700e+04, 7.168300e+04, 1.494300e+04, 2.347400e+04,
+        1.558000e+03, 2.903000e+03]
+    tensor_in_sizes = [1, 3, 3, 2]
+    filter_in_sizes = [2, 2, 2, 2]
+    bias_in_sizes = [2]
+
+    x = self._CreateNumpyTensor(tensor_in_sizes)
+    filter_in = self._CreateNumpyTensor(filter_in_sizes)
+    bias_in = self._CreateNumpyTensor(bias_in_sizes)
+    # To get different weights for filter
+    ofs = 1
+
+    conv1 = self._CreateConv2D(x, filter_in)
+    conv2 = self._CreateConv2D(conv1, filter_in + ofs)
+
+    conv = self._CreateConv2D(conv2, filter_in - ofs)
+    bias_add = nn_ops.bias_add(conv, bias_in)
+    add = math_ops.add_n([bias_add, conv1])
+
+    self.assertAllEqual(
+        np.rint(expected_output),
+        self.evaluate(add).reshape(-1))
+
+  @test_util.deprecated_graph_mode_only
+  def testAddWithRefCountTwoAndRunAddFirst(self):
```

Comment:
The comment here could be:
```
# Tests tensor forwarding of a fused Conv2D+BiasAdd+Add op when the input to Add
# has refcount 2 and Add (in the fused Conv2D op) is its first consumer.
```
Or we could use other op than Add in `output = math_ops.add_n([relu, conv2])` and remove the `(in the fused Conv2D op)` part in the comment.

Reply:
Done